File: miller_rabin_primality_test.cpp
/*
1. Write p-1 as 2^b * m where m is odd
2. Pick a random integer a such that 1 < a < p-1
3. Compute z = a^m mod p
4. If z == 1 or z == p-1, then p is probably prime
5. else,  For j = 1 to b-1, compute z = z^2 mod p
   - If z == p-1, then p is probably prime
   - If z == 1, then p is composite
   - if after b-1 iterations, z is not p-1 in any interation, then p is composite
6. Repeat the above steps for a certain number of iterations to increase confidence level


*/

#define ll long long
#include <bits/stdc++.h>
using namespace std;


ll mod_pow(ll a, ll d, ll mod)
{
    ll result = 1;
    while (d > 0)
    {
        if (d % 2 == 1)
            result = (result * a) % mod;

        a = (a * a) % mod;
        d /= 2;
    }
    return result;
}

bool millerRabin(ll p, int iterations = 10)
{
    if (p < 2)
        return false;
    if (p == 2 || p == 3)
        return true;
    if (p % 2 == 0)
        return false;

    // Step 1: p-1 = 2^b * m
    ll m = p - 1;
    int b = 0;

    while (m % 2 == 0)
    {
        m /= 2;
        b++;
    }

    for (int i = 0; i < iterations; i++)
    {
        // Step 2: pick a
        ll a = 2 + rand() % (p - 3);

        // Step 3: z = a^m mod p
        ll z = mod_pow(a, m, p);

        // Step 4: check
        if (z == 1 || z == p - 1)
            continue;

        bool passed = false;

        // Step 5: square up to b-1 times
        for (int j = 1; j < b; j++)
        {
            z = (z * z) % p;

            if (z == p - 1)
            {
                passed = true;
                break;
            }

            if (z == 1)
                return false; // composite
        }

        if (!passed)
            return false;
    }

    return true;
}

int main()
{
    ll p;
    cout << "Enter number: ";
    cin >> p;

    if (millerRabin(p, 20))
        cout << "Probably Prime\n";
    else
        cout << "Composite\n";

    return 0;
}

File: miller_rabin_primality_test.py
# write in python when need to work with big integer like 512 bit number


import random

def miller_rabin(p, iterations=20):

    # Step 0: handle small cases
    if p < 2:
        return False
    if p in (2, 3):
        return True
    if p % 2 == 0:
        return False

    # Step 1: write p-1 = 2^b * m
    m = p - 1
    b = 0

    while m % 2 == 0:
        m //= 2
        b += 1

    # Step 2: repeat test
    for _ in range(iterations):

        # pick a in [2, p-2]
        a = random.randint(2, p - 2)

        # Step 3: z = a^m mod p
        z = pow(a, m, p)

        # Step 4: check
        if z == 1 or z == p - 1:
            continue

        # Step 5: square up to b-1 times
        passed = False

        for _ in range(b - 1):
            z = pow(z, 2, p)

            if z == p - 1:
                passed = True
                break

            if z == 1:
                return False  # composite

        if not passed:
            return False

    return True



n = random.getrandbits(512) # take a random 512 bit number
n |= 1  # make it odd
# n= 13407807929942597099574024998205846127479365820592393377723561443721764030073546976801874298166903427690031858186486050853753882811946569946433649006084171

print("Generated Number:\n")
print(n)

if miller_rabin(n, 20):
    print("Probably Prime")
else:
    print("Composite")